src/olympia/versions/tasks.py

```diff
@@ -126,11 +211,29 @@ def extract_version_to_git(version_id, author_id=None, note=None):
     log.info('Extracting {version_id} into git backend'.format(
         version_id=version_id))
 
-    repo = AddonGitRepository.extract_and_commit_from_version(
-        version=version, author=author, note=note)
+    try:
+        repo = AddonGitRepository.extract_and_commit_from_version(
+            version=version, author=author, note=note)
+
+        log.info('Extracted {version} into {git_path}'.format(
+            version=version_id, git_path=repo.git_repository_path))
+    except BrokenRefError as err:
+        # This is needed to prevent a potential infinite loop if a broken
+        # reference is detected in `extract_addon_to_git()`, which we call
+        # later in this block.
+        if stop_on_broken_ref:
+            raise err
 
-    log.info('Extracted {version} into {git_path}'.format(
-        version=version_id, git_path=repo.git_repository_path))
+        # Reset the git hash of each version of the add-on related to the
+        # current add-on because we want to re-extract everything.
+        addon.versions.update(git_hash='')
+        # Retrieve the repo for the add-on and delete it.
+        addon_repo = AddonGitRepository(addon, package_type='addon')
+        addon_repo.delete()
```

Comment:
I think you should add the lock on the addon before you do this. Otherwise, if there is *another* extraction in progress for that addon that is somehow not broken (maybe because it's on a different, not broken channel) it might go through while you're deleting the repos...

In fact, I think the lock should be set super early in all extraction tasks (addon and version) and released once they are done, in order to prevent parallel extractions of multiple versions of a given add-on.

You'll also need something to remove the lock in case anything goes wrong in all tasks, otherwise you might end up locking an add-on for all eternity...

Reply:
> I think you should add the lock on the addon before you do this. Otherwise, if there is _another_ extraction in progress for that addon that is somehow not broken (maybe because it's on a different, not broken channel) it might go through while you're deleting the repos...

Ah, right. That would make sense but then we'll need to pass a `can_be_delayed=False` param to the `extract_addon_to_git` task. I tried to keep the lock/unlock in the same task so that if it fails, the add-on is still unlocked at the end.

Locking in this task and unlocking in another task is going to be... interesting.

Reply:
> Ah, right. That would make sense but then we'll need to pass a `can_be_delayed=False` param to the `extract_addon_to_git` task. I tried to keep the lock/unlock in the same task so that if it fails, the add-on is still unlocked at the end.

Yes, and also... when you trigger the task to rebuild the repos entirely for the addon, you shouldn't obey the lock that was set for the extraction of the version in which we noticed the brokenness... But at the same time, you can't release the lock too early (you don't want other version extractions to go through) and you can't completely ignore locks in the rebuild task either (cause you don't want multiple rebuilds in parallel). Maybe you need a second lock mechanism: one that says "a version is being extracted" and a different one that says "the whole addon repos is being rebuilt"... That all depends on how you implement triggering the extractions in the rebuild. 
> 
> Locking in this task and unlocking in another task is going to be... interesting.

Reply:
Yeah, I fixed the easy things so now I am looking into that.

What if we don't delete the repo in the `extract_version_to_git` task and pass `delete_repo=True` to the `extract_addon_to_git` task? Wouldn't that solve our problem here?

Reply:
It'd certainly be a good thing. But we probably still want the extract version task to not do anything while a full addon extraction (or another version extraction) is underway...

Reply:
So there are two things here:

> we probably still want the extract version task to not do anything while a full addon extraction is underway...

This is why we have the `git_extraction_is_in_progress` lock, no? It is set before we start the full add-on extraction and it is checked in the version task.

> we probably still want the extract version task to not do anything while a version extraction is underway...

That is different. We do not have that right now.